### PR TITLE
test: e2e - cypress notification rules

### PR DESCRIPTION
**Pull request by karel-rehor**
Adds additional e2e cypress tests for notification rules

   * Filter notification rules list
   * Notification list card
       * Change name
       * Change description
       * Activate / Deactivate
       * Verify existence and sanity of other field values 
   * Cone notification rule
      * Modify values of clone   
   * Delete notification rule

### CODE REVIEW

**Comment by ui-ci-bot-influx** [comment]
Metric | Score | Level
------------ | ------------- | ------------- 
performance | 50 | :red_circle:
accessibility | 97 | :green_circle:
seo | 90 | :green_circle:
pwa | null | :hankey:
best-practices | 100 | :green_circle:


[Compare your scores with master branch](https://influxdb.aws.influxdata.io/orgs/844910ece80be8bc/dashboards/073c4b95f1089000?lower=now%28%29%20-%202d&vars%5Blighthouse_score_branch%5D=3999)

File: cypress/e2e/shared/notificationRules.test.ts [hunk 1]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+    beforeEach(() => {
+      cy.writeLPDataFromFile({
+        filename: 'data/wumpus01.lp',
+        offset: '20m',
+        stagger: '1m',
+      })
+      cy.get<Organization>('@org').then((org: Organization) => {
+        cy.get<Bucket>('@bucket').then((bucket: Bucket) => {
+          // get default org and bucket
+          // 2. create check
```
Comment (wiedld) [inline cypress/e2e/shared/notificationRules.test.ts:803]: cy.log() creating check.

File: cypress/e2e/shared/notificationRules.test.ts [hunk 2]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+              builderConfig: {
+                ...check.query.builderConfig,
+                buckets: [bucket.id],
+              },
+            },
+          }).then((resp: {body: any}) => {
+            cy.wrap(resp.body).as('check')
+          })
+
+          cy.createEndpoint({
```
Comment (wiedld) [inline cypress/e2e/shared/notificationRules.test.ts:820]: cy.log() creating endpoint.

File: cypress/e2e/shared/notificationRules.test.ts [hunk 3]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+            ...(endp as NotificationEndpoint),
+            orgID: org.id,
+          }).then(resp => {
+            cy.wrap(resp.body).as('endp')
+          })
+
+          cy.get<NotificationEndpoint>('@endp').then(endp => {
```
Comment (wiedld) [inline cypress/e2e/shared/notificationRules.test.ts:827]: cy.log() creating rule.

File: cypress/e2e/shared/notificationRules.test.ts [hunk 4]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+          expect(dialog.text()).to.match(/^Last Run Status:.*success.*/)
+        })
+    })
+  })
+
+  describe('Cloning', () => {
+    const rule2Clone = {
+      name: 'Despair',
+      description:
+        'Nabokov.  Hermann Karlovich, a Russian of German descent and owner of a chocolate factory, meets a homeless man in the city of Prague, who he believes is his doppelgänger.',
```
Comment (wiedld) [inline cypress/e2e/shared/notificationRules.test.ts:1016]: thanks for wikipedia detour. 😆

File: cypress/e2e/shared/notificationRules.test.ts [hunk 5]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+
+          cy.createEndpoint({
+            ...(endp as NotificationEndpoint),
+            orgID: org.id,
+          }).then(resp => {
+            cy.wrap(resp.body).as('endp')
+          })
+
+          cy.get<NotificationEndpoint>('@endp').then(endp => {
+            cy.createRule({
```
Comment (wiedld) [inline cypress/e2e/shared/notificationRules.test.ts:1059]: the repeated uses of cy.createCheck, cy.createEndpoint, and cy.createRule suggests some helped functions may make this easier. (plus easier to build future tests).

Not blocking for the PR. Just a suggestion.

Comment (wiedld) [reply cypress/e2e/shared/notificationRules.test.ts:1059]: if you don't extract to helper functions, then please make sure to add cy.logs.

File: cypress/e2e/shared/notificationRules.test.ts [hunk 6]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+          cy.wrap(
+            elem
+              .text()
+              .replace('Task ID: ', '')
+              .replace('Copy to Clipboard', '')
+          ).as('TargetTaskID')
+        })
+      })
+
+      // verify basic identifying parameters
```
Comment (wiedld) [inline cypress/e2e/shared/notificationRules.test.ts:1098]: same pattern. cy.log() this comment.

File: cypress/e2e/shared/notificationRules.test.ts [hunk 7]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+        })
+        cy.getByTestID(`label--pill ${rule2Clone.label}`).should('exist')
+        cy.getByTestID('resource-list--editable-description').should(
+          'have.text',
+          rule2Clone.description
+        )
+        cy.getByTestID('rule-card--name').click()
+      })
+
+      // verify then change cloned values
```
Comment (wiedld) [inline cypress/e2e/shared/notificationRules.test.ts:1122]: same pattern. cy.log() this comment.

File: cypress/e2e/shared/notificationRules.test.ts (hunk 6 above)
Comment (wiedld) [reply cypress/e2e/shared/notificationRules.test.ts:1098]: but it looks like repeated code (capturing IDs). can alternatively decide on helper functions.

File: cypress/e2e/shared/notificationRules.test.ts [hunk 8]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+        .should('have.text', rule.statusRules[0].currentLevel as string)
+        .click()
+      cy.getByTestID(`levels--dropdown-item ${newLevel}`).click()
+
+      cy.getByTestID('endpoint--dropdown--button').contains(endp.name)
+
+      cy.getByTestID('rule-overlay-save--button').click()
+
+      cy.wait('@putClonedRule').then(({request: {body}}) => {
+        // verify internal values not displayed e.g. every, offset, level etc
```
Comment (wiedld) [inline cypress/e2e/shared/notificationRules.test.ts:1159]: same pattern. cy.log() this comment.

File: cypress/e2e/shared/notificationRules.test.ts [hunk 9]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+        expect(body.every).to.equal(newEvery)
+        expect(body.offset).to.equal(newOffset)
+        expect(body.statusRules[0].currentLevel).to.equal(newLevel)
+        cy.get('@TargetTaskID').then(taskId => {
+          expect(body.taskID).to.not.equal(taskId)
+        })
+
+        // then verify changed display value - name
```
Comment (wiedld) [inline cypress/e2e/shared/notificationRules.test.ts:1167]: same pattern. cy.log() this comment.

File: cypress/e2e/shared/notificationRules.test.ts [hunk 10]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+        cy.getByTestID(`rule-card ${newName}`)
+          .within(() => {
+            cy.getByTestID('rule-card--name').should('have.text', newName)
+
+            // and delete the rule
```
Comment (wiedld) [inline cypress/e2e/shared/notificationRules.test.ts:1172]: same pattern. cy.log() this comment.

**Review by wiedld — CHANGES_REQUESTED** [review]
The vast majority of my requests are about converting code comments into cy logs.
Then two lines of selector changes.

And a (non-blocking) request to use helper methods.

File: cypress/e2e/shared/notificationRules.test.ts [hunk 11]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+                })
+              })
+            })
+          })
+        })
+      })
+    })
+
+    it('can filter the rule list', () => {
+      // baseline check
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Done

File: cypress/e2e/shared/notificationRules.test.ts [hunk 12]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+      cy.getByTestID('rule-card--name')
+        .should('have.length', 4)
+        .then(names => {
+          names.toArray().forEach((name: HTMLElement, index: number) => {
+            expect(name.innerText).to.equal(Rules[index].name)
+          })
+        })
+
+      // filter name
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Done

File: cypress/e2e/shared/notificationRules.test.ts [hunk 13]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+      cy.getByTestID('filter--input rules').type(Rules[1].name.substr(0, 5))
+
+      cy.getByTestID('rule-card--name')
+        .should('have.length', 1)
+        .should('have.text', Rules[1].name)
+
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts:856]: Done

File: cypress/e2e/shared/notificationRules.test.ts [hunk 14]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+        .should('have.length', 4)
+        .then(names => {
+          names.toArray().forEach((name: HTMLElement, index: number) => {
+            expect(name.innerText).to.equal(Rules[index].name)
+          })
+        })
+    })
+
+    it('can modify a notification rule entry', () => {
+      cy.log('modifying notification rule entry')
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Removed

File: cypress/e2e/shared/notificationRules.test.ts [hunk 15]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+            ).as('TargetTaskID')
+          })
+
+          cy.get<string>('@TargetRuleID').then(id => {
+            cy.intercept('PATCH', `/api/v2/notificationRules/${id}`).as(
+              'patchNotificationRule'
+            )
+          })
+
+          // modify description
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Done

File: cypress/e2e/shared/notificationRules.test.ts [hunk 16]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+          cy.getByTestID('input-field')
+            .clear()
+            .type(`${newRule.description}{enter}`)
+          cy.wait('@patchNotificationRule')
+          cy.getByTestID('resource-list--editable-description').should(
+            'have.text',
+            newRule.description
+          )
+
+          // modify name
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Done

File: cypress/e2e/shared/notificationRules.test.ts [hunk 17]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+          })
+        })
+    })
+
+    it('can modify a notification rule entry', () => {
+      cy.log('modifying notification rule entry')
+      // baseline asserts
+      cy.getByTestID(`rule-card ${Rules[1].name}`)
+        .within(() => {
+          cy.getByTestID('rule-card--name').should('have.text', Rules[1].name)
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts:873]: Done

File: cypress/e2e/shared/notificationRules.test.ts [hunk 18]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+          cy.getByTestID('rule-card--name-button').click()
+          cy.getByTestID('rule-card--input').type(`${newRule.name}{enter}`)
+        })
+        .then(() => {
+          // When name is modified sort order of cards should change
+          Rules[1] = newRule
+          Rules = Rules.sort(rulesSort)
+          cy.wait('@patchNotificationRule')
+          cy.getByTestID(`rule-card ${newRule.name}`).should('exist')
+          // Verify New Sort Order and card name change
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Done

File: cypress/e2e/shared/notificationRules.test.ts [hunk 19]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+          cy.getByTestID('rule-card--name')
+            .should('have.length', 4)
+            .then(names => {
+              names.toArray().forEach((name: HTMLElement, index: number) => {
+                expect(name.innerText).to.equal(Rules[index].name)
+              })
+            })
+
+          // Verify task ID is the same
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Done

File: cypress/e2e/shared/notificationRules.test.ts [hunk 20]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+          cy.getByTestID(`rule-card ${newRule.name}`).within(() => {
+            cy.get('@TargetTaskID').then(id => {
+              cy.getByTestID('copy-task-id').should('contain.text', id)
+            })
+
+            // check sanity of update and run times
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Done

File: cypress/e2e/shared/notificationRules.test.ts [hunk 21]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+            cy.getByTestID('resource-list--meta').within(() => {
+              const now = new Date()
+              cy.get('> div')
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Done

File: cypress/e2e/shared/notificationRules.test.ts [hunk 22]
```diff
@@ -757,9 +757,426 @@ describe('NotificationRules', () => {
...
+                .then(elem => {
+                  expect(parseInt(elem.text().replace(/\D+/g, '')))
+                    .to.be.at.least(0)
+                    .and.to.be.at.most(Cypress.config('requestTimeout') / 1000)
+                })
+            })
+          })
+        })
+
+      // disable / enable
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Done

File: cypress/e2e/shared/notificationRules.test.ts (hunk 2 above)
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Done

File: cypress/e2e/shared/notificationRules.test.ts (hunk 3 above)
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Done

File: cypress/e2e/shared/notificationRules.test.ts (hunk 4 above)
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts:988]: No problem

File: cypress/e2e/shared/notificationRules.test.ts (hunk 5 above)
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Added a helper command to support/commands.ts

File: cypress/e2e/shared/notificationRules.test.ts (hunk 6 above)
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: Added cy.log, but left off creating a helper function for capturing IDs. Did not see how a helper function would result in less code than the currently implementation using cypress/

File: cypress/e2e/shared/notificationRules.test.ts (hunk 7 above)
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: done

File: cypress/e2e/shared/notificationRules.test.ts (hunk 8 above)
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: done

File: cypress/e2e/shared/notificationRules.test.ts (hunk 9 above)
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: done

File: cypress/e2e/shared/notificationRules.test.ts (hunk 10 above)
Comment (karel-rehor) [reply cypress/e2e/shared/notificationRules.test.ts]: done

**Comment by karel-rehor** [comment]
Hi Denise, 

I've added the changes, but it appears there's been an outage at Circle/CI, so waiting on the results. 

https://status.circleci.com/ 
https://app.circleci.com/pipelines/github/influxdata/ui?branch=pull%2F3999&filter=all

Unless of course the UI tests have already been removed from the build pipeline.  I heard that their removal might be coming. 

-K

File: cypress/support/commands.ts [hunk 23]
```diff
@@ -910,6 +912,57 @@ export const createRule = (
   return cy.request('POST', 'api/v2/notificationRules', rule)
 }
 
+// alertGroup
+export const createAlertGroup = (
```
Comment (wiedld) [inline cypress/support/commands.ts:916]: 👍🏼

**Review by wiedld — APPROVED** [review]
Nice changes.

**Pull request merged**